spec/requests/spree/admin/payments_spec.rb

```diff
@@ -0,0 +1,135 @@
+    sign_in create(:admin_user)
+  end
+
+  describe "PUT /admin/orders/:order_number/payments/:id/fire" do
+    let(:payment) do
+      create(
+        :payment,
+        payment_method: stripe_payment_method,
+        order:,
+        response_code: "pi_123",
+        amount: order.total,
+        state: "completed"
+      )
+    end
+    let(:stripe_payment_method) do
+      create(:stripe_sca_payment_method, distributors: [order.distributor])
+    end
+    let(:headers) { { HTTP_REFERER: spree.admin_order_payments_url(order) } }
+
+    before do
+      order.update(payments: [])
+      order.payments << payment
+    end
+
+    context "with no event parameter" do
+      it "redirect to payments page" do
+        put(
+          "/admin/orders/#{order.number}/payments/#{order.payments.first.id}/fire",
+          params: {},
+          headers:
+        )
+
+        expect(response).to redirect_to(spree.admin_order_payments_url(order))
+      end
+    end
+
+    context "with no payment source" do
+      it "redirect to payments page" do
+        allow(payment).to receive(:payment_source).and_return(nil)
+
+        put(
+          "/admin/orders/#{order.number}/payments/#{order.payments.first.id}/fire?e=void",
+          params: {},
+          headers:
+        )
+
+        expect(response).to redirect_to(spree.admin_order_payments_url(order))
+      end
+    end
+
+    context "with 'void' parameter" do
+      before do
+        allow(Spree::Payment).to receive(:find).and_return(payment)
+      end
+
+      it "calls void_transaction! on payment" do
+        expect(payment).to receive(:void_transaction!)
+
+        put(
+          "/admin/orders/#{order.number}/payments/#{order.payments.first.id}/fire?e=void",
+          params: {},
+          headers:
+        )
+      end
+
+      it "redirect to payments page" do
+        allow(payment).to receive(:void_transaction!).and_return(true)
+
+        put(
+          "/admin/orders/#{order.number}/payments/#{order.payments.first.id}/fire?e=void",
+          params: {},
+          headers:
+        )
+
+        expect(response).to redirect_to(spree.admin_order_payments_url(order))
+        expect(flash[:success]).to eq "Payment Updated"
+      end
+
+      context "when void_transaction! fails" do
+        it "set an error flash message" do
+          allow(payment).to receive(:void_transaction!).and_return(false)
+
+          put(
+            "/admin/orders/#{order.number}/payments/#{order.payments.first.id}/fire?e=void",
+            params: {},
+            headers:
+          )
+
+          expect(response).to redirect_to(spree.admin_order_payments_url(order))
+          expect(flash[:error]).to eq "Could not update the payment"
+        end
+      end
+    end
+
+    context "when something unexpected happen" do
+      before do
+        allow(Spree::Payment).to receive(:find).and_return(payment)
+        allow(payment).to receive(:void_transaction!).and_raise(StandardError, "Unexpected !")
```

Comment:
We could use an exception raised by the gateway, but it's also printed in the logs somewhere else.  I added this mainly to cover scenario we are not anticipating. Ie, when I first looked into this bug I couldn't see what was happening because the error happened in the Stripe decorator and was getting swallowed. We could run into this again if says something break in active merchant, so I though it was better not use a "real example"  ( as in an exception we can reasonably expect to happen sometimes ).

That makes me think, shouldn't we track this in Bugsnag ? If I am no mistaken, because the exception is handled in the code, it won't be picked up by Bugsnag. Though ? ping @mkllnk

Reply:
True, if we rescue the error then it won't appear in Bugsnag. And if we are not anticipating this then it would be good to notify Bugsnag, just in case.